google/cloud/bigquery/query.py

```diff
@@ -119,6 +120,18 @@ def to_api_repr(self):
         # attributes in the API representation when needed. Here we omit them.
         return {"type": self._type}
 
+    def with_name(self, new_name: Optional[str]):
```

Comment:
Why make name optional?

Reply:
We allow explicit `None`, the same as the constructor does (to allow clearing the name). Perhaps using the alternative form `Union[str, None]` would make it less confusing, as the argument itself is not optional.

Reply:
You're right that the argument is required.  But I still don't see a point in allowing `None`.  Nameless types are readily available in `SqlParameterScalarTypes`.  To expect someone to call `with_name` and, you know, not provide a name seems silly. :)  

I don't think changing `Optional[str]` to `Union[str, None]` is an improvement.

I'm willing to defer.

Reply:
I tend to not assume about the users' use cases too much, as several times in the past we have gotten that wrong. :)

I don't feel strongly about either way (supporting or not supporting clearing the name), but since the maintenance burden is more or less the same for both options, supporting `None` seems fine here.